Add evalf to ImmutableDenseNDimArray

jgonzalez49 opened the pull request:
#### Brief description of what is fixed or changed
Add evalf to ImmutableDenseNDimArray.

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* tensor
  * Add evalf to ImmutableDenseNDimArray.
<!-- END RELEASE NOTES -->

oscarbenjamin commented on sympy/tensor/array/dense_ndim_array.py:
```diff
@@ -157,6 +157,10 @@ def _eval_simplify(self, **kwargs):
         from sympy.simplify.simplify import simplify
         return self.applyfunc(simplify)
 
+    def evalf(self, **kwargs):
+        f=lambda x: x.evalf()
+        return self.applyfunc(f)
```
Rather than defining `evalf` then `EvalfMixin` class should be subclassed somewhere.

jgonzalez49 replied:
> Rather than defining `evalf` then `EvalfMixin` class should be subclassed somewhere.

Hi, thanks for the feedback. As I'm not familiar with this code base I'm doing a bit of digging based on your suggestion, but if you have an example in mind for me to review, I can attempt to match that. I did find this example in the matrices section that is similar to my original implementation, but supports more options than my patch:

https://github.com/sympy/sympy/blob/master/sympy/matrices/common.py#L2027

oscarbenjamin replied:
There are several examples:
```console
$ git grep EvalfMixin
doc/src/contributing/dependencies.md:  [evalf](sympy.core.evalf.EvalfMixin.evalf).
doc/src/explanation/glossary.md:    [*`evalf`*](sympy.core.evalf.EvalfMixin.evalf) is the method present on
doc/src/explanation/glossary.md:    <sympy.core.evalf.EvalfMixin.n>` method and {func}`~.N` function are both
doc/src/guides/assumptions.rst:number with :py:meth:`~.EvalfMixin.evalf`:
doc/src/guides/assumptions.rst:call to :py:meth:`~.EvalfMixin.evalf` will fully evaluate. The
doc/src/modules/core.rst:.. autoclass:: EvalfMixin
sympy/core/basic.py:        sympy.core.evalf.EvalfMixin.evalf: calculates the given formula to a desired level of precision
sympy/core/evalf.py:        ``EvalfMixin.evalf`` and in addition,
sympy/core/evalf.py:class EvalfMixin:
sympy/core/expr.py:from .evalf import EvalfMixin, pure_complex, DEFAULT_MAXPREC
sympy/core/expr.py:class Expr(Basic, EvalfMixin):
sympy/core/relational.py:from .evalf import EvalfMixin
sympy/core/relational.py:class Relational(Boolean, EvalfMixin):
sympy/geometry/entity.py:from sympy.core.evalf import EvalfMixin, N
sympy/geometry/entity.py:class GeometryEntity(Basic, EvalfMixin):
sympy/geometry/tests/test_entity.py:def test_geometry_EvalfMixin():
sympy/physics/control/lti.py:from sympy.core.evalf import EvalfMixin
sympy/physics/control/lti.py:class LinearTimeInvariant(Basic, EvalfMixin):
sympy/physics/vector/dyadic.py:from sympy.core.evalf import EvalfMixin
sympy/physics/vector/dyadic.py:class Dyadic(Printable, EvalfMixin):
sympy/physics/vector/vector.py:from sympy.core.evalf import EvalfMixin
sympy/physics/vector/vector.py:class Vector(Printable, EvalfMixin):
sympy/sets/sets.py:from sympy.core.evalf import EvalfMixin
sympy/sets/sets.py:class Set(Basic, EvalfMixin):
```

jgonzalez49 committed:
Update ndim_array.py

Added EvalfMixin class as a subclass and merged in matrix evalf code as _eval_evalf.

jgonzalez49 commented:
Following the way EvalfMixin was used in the vector class, and recycling what was already done in the matrix class, I have updated my patch. @oscarbenjamin Please let me know if there is a better way for me to be in line with what you were suggesting.

jgonzalez49 committed:
Update .mailmap

Added name to mailmap.

jgonzalez49 committed:
Merge branch 'sympy:master' into master

jgonzalez49 committed:
Update .mailmap

Updated name address in mailmap.

jgonzalez49 committed:
Update .mailmap

Added other name address to mailmap.

1e9abhi1e10 commented on .mailmap:
```diff
@@ -729,12 +729,14 @@ Jennifer White <jcrw122@googlemail.com>
 Jens H. Nielsen <jenshnielsen@gmail.com>
 Jens Jørgen Mortensen <jj@smoerhul.dk>
 Jeremey Gluck <jeremygluck@yahoo.com>
+Jeremias Gonzalez <jgonzalez49@ucmerced.edu>
```
```suggestion
Jeremias Gonzalez <jgonzalez49@ucmerced.edu> jgonzalez49 <68764159+jgonzalez49@users.noreply.github.com>
```

1e9abhi1e10 commented on .mailmap:
```diff
@@ -729,12 +729,14 @@ Jennifer White <jcrw122@googlemail.com>
 Jerry James <loganjerry@gmail.com>
 Jerry Li <jerry@jerryli.ca>
 Jezreel Ng <jezreel@gmail.com>
+jgonzalez49 <68764159+jgonzalez49@users.noreply.github.com>
```
Remove this one.

jgonzalez49 committed:
Update .mailmap

Reordered mailmap.

jgonzalez49 committed:
Update .mailmap

Merged other address to one name.

jgonzalez49 committed:
Update .mailmap

Added original name to line.

jgonzalez49 commented:
> It appears that this is your initial pull request, so you should include your name and email in the `.mailmap` file. https://docs.sympy.org/dev/contributing/new-contributors-guide/workflow-process.html#add-your-name-and-email-address-to-the-mailmap-file

@1e9abhi1e10 Thanks for letting me know, I have incorporated your code comments into the current commit.

oscarbenjamin commented on sympy/tensor/array/ndim_array.py:
```diff
@@ -586,6 +587,12 @@ def _check_index_for_getitem(self, index):
 
         return index
 
+    def _eval_evalf(self, n=15, subs=None, maxn=100, chop=False, strict=False, quad=None, verbose=False):
+        """Apply evalf() to each element of self."""
+        options = {'subs':subs, 'maxn':maxn, 'chop':chop, 'strict':strict,
+                'quad':quad, 'verbose':verbose}
+        return self.applyfunc(lambda i: i.evalf(n, **options))
```
In this situation it would be better for the method to take `**options` and just pass them on rather than listing out all of the arguments explicitly.

jgonzalez49 replied:
@oscarbenjamin I believe the new commit addresses this. Also, since the code originally came from the matrix class, let me know if I should open a parallel pull request to simplify that part of the code the same way.

oscarbenjamin replied:
That could also be fixed. Actually looking at `git grep _eval_evalf` it does not look like `_eval_evalf` should take any arguments apart from `prec`.

jgonzalez49 committed:
Update ndim_array.py

Pass through options.

oscarbenjamin commented on sympy/tensor/array/ndim_array.py:
```diff
@@ -586,6 +587,10 @@ def _check_index_for_getitem(self, index):
 
+    def _eval_evalf(self, **options):
+        """Apply evalf() to each element of self."""
+        return self.applyfunc(lambda i: i.evalf(n, **options))
```
I guess both `*args` and `**kwargs` need to be used to make this work.

jgonzalez49 replied:
Oops, that's my mistake for not replacing all the original options (I missed the n). I have updated it now to make use of the args and kwargs parameters.

jgonzalez49 committed:
Update ndim_array.py

Corrected passing of arguments.

oscarbenjamin commented:
Looking closer it should perhaps be:
```python
def _eval_evalf(self, prec):
    dps = prec_to_dps(prec)
    return self.applyfunc(lambda e: e.evalf(dps))
```

jgonzalez49 committed:
Update common.py

Simplifying arguments for evalf in matrix class.

jgonzalez49 committed:
Update ndim_array.py

Simplifying arguments to only prec.

jgonzalez49 commented:
I've simplified the arguments for both the matrix and array classes to your suggested code.

jgonzalez49 committed:
Update ndim_array.py

Added import for prec_to_dps.

jgonzalez49 committed:
Update common.py

Added import for prec_to_dps.

jgonzalez49 committed:
Update common.py

Add default for prec argument.

jgonzalez49 committed:
Update ndim_array.py

Add default for prec argument.

jgonzalez49 committed:
Update common.py

Removed changes from earlier commit.

jgonzalez49 commented:
I removed the changes I made to the matrix class because of multiple test failures depending on the keywords being passed in the original code.

oscarbenjamin commented on sympy/tensor/array/ndim_array.py:
```diff
@@ -586,6 +588,11 @@ def _check_index_for_getitem(self, index):
 
         return index
 
+    def _eval_evalf(self, prec=15):
```
The default here should be `53` rather than `15` but actually it is just better not to provide any default because a default is not needed.

jgonzalez49 replied:
The default has been removed. It was from an attempt to mitigate the errors earlier from the matrix class changes (and have parity between the matrix and array classes), but now it's back to the original suggested code.

oscarbenjamin commented:
> I removed the changes I made to the matrix class because of multiple test failures depending on the keywords being passed in the original code.

Can you give an example? It should be possible to make the same change for `Matrix` or otherwise this change might not be right for `Array`...

jgonzalez49 commented:
> Can you give an example? It should be possible to make the same change for `Matrix` or otherwise this change might not be right for `Array`...

Here's the last round of tests that were run against the simplified arguments for the matrix code:

https://github.com/sympy/sympy/actions/runs/6178490381

I'm not sure about the cause for all of them failing, but at least one easy example is from tests-latest (2) where the test was passing a keyword `n` but that no longer would be included as a keyword with the simplified code:

https://github.com/sympy/sympy/actions/runs/6178490381/job/16771838997#step:6:414

Other failures like this one I'm not sure about:

https://github.com/sympy/sympy/actions/runs/6178490381/job/16771838450

oscarbenjamin commented:
The n vs prec thing is because of `.evalf()` vs `._eval_evalf()`. It's really unclear which method should be defined when using `EvalFMixin`...

jgonzalez49 committed:
Update ndim_array.py
